packages/lodestar/src/chain/forkChoice/statefulDag/lmdGhost.ts

```diff
@@ -216,8 +226,10 @@ export class StatefulDagLMDGHOST implements ILMDGHOST {
   }
 
   public setJustified(blockRoot: Hash): void {
-    const rootHex = blockRoot.toString("hex");
-    this.justified = this.nodes[rootHex];
+    if (this.shouldUpdateJustifiedCheckpoint(blockRoot)) {
```

Comment:
Aren't we already doing this check in chain before calling this method?

Reply:
No, this is new method and specific for fork choice.

Reply:
This could leave us in wierd state where we would update database and everything while we weren't actually justified.

As it seems we have two options:
1. change this method to return "true" if it updated fork choice - in chain module call this method first and update database and other stuff
2. Leave fork choice as is and call `shouldUpdateJustifiedCheckpoint` in chain module

I'm leaning on option 1 as I think chain module should follow fork choice decisions
 
cc @wemeetagain

Reply:
yeah I see the concern. From the spec of https://github.com/ethereum/eth2.0-specs/blob/dev/specs/core/0_fork-choice.md, I only see we use this new method in the "`store`".

Reply:
Actually on second look I might be wrong. Chain updates database depending on if state transition justified epoch (based on attestations). Still not sure what happens if state marks epoch as justified but fork choice doesn't set it.Might be irrelevant. It would be awesome if cayman can explain this better just to make it clear

Reply:
hmm. ok, so something is a little funky here, didn't catch it the first time around.
`forkChoice.setJustified` is currently only being called on an epoch boundary, once the chain sees a new justified epoch (src/chain/chain.ts#L356). This 'bounce' attack is meant to not update justified blocks in later slots within an epoch, but this is was never a case that was considered in our implementation, because we only update on epoch boundaries anyways.
I think we may want to take a step back at what we're currently doing, and maybe revamp the whole process.
It doesn't even look like we're applying the fork choice at all (src/chain/chain.ts#L338)

Reply:
The issue to me regarding understanding the spec is that there are spec for fork choice and spec for beacon chain but they do not link together, so it's not clear to me on when to set justified check point to `forkChoice/store` and when to call `get_head`. 

The other thing to consider is we have a different implementation compared to the spec. Should we update justified check point to fo forkChoice like this in the spec
```
if state.current_justified_checkpoint.epoch > store.justified_checkpoint.epoch:
        store.best_justified_checkpoint = state.current_justified_checkpoint
        if should_update_justified_checkpoint(store, state.current_justified_checkpoint):
            store.justified_checkpoint = state.current_justified_checkpoint
```
or we do that on epoch boundary like in the current implementation.